File: sdk/java/src/main/java/vn/hitechcloud/sdk/HiTechCloudClient.java
package vn.hitechcloud.sdk;

import vn.hitechcloud.sdk.resources.*;

import java.util.Map;

public class HiTechCloudClient {
    public static final String PRODUCTION_URL = "https://api.hitechcloud.vn";
    public static final String OTE_URL = "https://ote.hitechcloud.vn";
    public static final String VERSION = "1.0.0";

    private final HttpClient httpClient;
    private final Auth auth;
    private final Users users;
    private final Affiliate affiliate;
    private final Billing billing;
    private final Support support;
    private final Contacts contacts;
    private final Domains domains;
    private final DnsManage dnsManage;
    private final Ssl ssl;
    private final Services services;
    private final Cart cart;
    private final Dns dns;
    private final Notifications notifications;
    private final Virtualizor virtualizor;
    private final CloudGpu cloudGpu;
    private final CloudService cloudService;
    private final NetworkServices networkServices;
    private final ServiceStatus serviceStatus;
    private final BareMetal bareMetal;
    private final Collocation collocation;
    private final Vcloudstack vcloudstack;
    private final Hosting hosting;
    private final LocationV2 locationV2;
    private final CloudInstance cloudInstance;
    private final CloudVm cloudVm;
    private final AffiliatesAdvanced affiliatesAdvanced;
    private final AiFactory aiFactory;
    private final PasskeyV2 passkeyV2;
    private final EmailMfaV2 emailMfaV2;
    private final Vneidekyc vneidekyc;
    private final WillExpired willExpired;
    private final UrlShortener urlShortener;
    private final ProxmoxBackup proxmoxBackup;
    private final CephS3 cephS3;
    private final Pmg pmg;
    private final Proxmox proxmox;
    private final Ipam ipam;
    private final Partner partner;

    public HiTechCloudClient(String token) {
        this(token, PRODUCTION_URL, 30, 3);
    }

    public HiTechCloudClient(String token, String baseUrl, int timeout, int maxRetries) {
        this.httpClient = new HttpClient(baseUrl, timeout, maxRetries);
        if (token != null) this.httpClient.setToken(token);

        this.auth = new Auth(httpClient);
        this.users = new Users(httpClient);
        this.affiliate = new Affiliate(httpClient);
        this.billing = new Billing(httpClient);
        this.support = new Support(httpClient);
        this.contacts = new Contacts(httpClient);
        this.domains = new Domains(httpClient);
        this.dnsManage = new DnsManage(httpClient);
        this.ssl = new Ssl(httpClient);
        this.services = new Services(httpClient);
        this.cart = new Cart(httpClient);
        this.dns = new Dns(httpClient);
        this.notifications = new Notifications(httpClient);
        this.virtualizor = new Virtualizor(httpClient);
        this.cloudGpu = new CloudGpu(httpClient);
        this.cloudService = new CloudService(httpClient);
        this.networkServices = new NetworkServices(httpClient);
        this.serviceStatus = new ServiceStatus(httpClient);
        this.bareMetal = new BareMetal(httpClient);
        this.collocation = new Collocation(httpClient);
        this.vcloudstack = new Vcloudstack(httpClient);
        this.hosting = new Hosting(httpClient);
        this.locationV2 = new LocationV2(httpClient);
        this.cloudInstance = new CloudInstance(httpClient);
        this.cloudVm = new CloudVm(httpClient);
        this.affiliatesAdvanced = new AffiliatesAdvanced(httpClient);
        this.aiFactory = new AiFactory(httpClient);
        this.passkeyV2 = new PasskeyV2(httpClient);
        this.emailMfaV2 = new EmailMfaV2(httpClient);
        this.vneidekyc = new Vneidekyc(httpClient);
        this.willExpired = new WillExpired(httpClient);
        this.urlShortener = new UrlShortener(httpClient);
        this.proxmoxBackup = new ProxmoxBackup(httpClient);
        this.cephS3 = new CephS3(httpClient);
        this.pmg = new Pmg(httpClient);
        this.proxmox = new Proxmox(httpClient);
        this.ipam = new Ipam(httpClient);
        this.partner = new Partner(httpClient);
    }

    public Map<String, Object> login(String email, String password) throws Exception {
        Map<String, Object> result = auth.login(email, password);
        if (result != null && result.containsKey("token")) {
            httpClient.setToken(result.get("token").toString());
        }
        return result;
    }

    public Map<String, Object> logout() throws Exception {
        Map<String, Object> result = auth.logout();
        httpClient.clearToken();
        return result;
    }

    public Auth auth() { return auth; }
    public Users users() { return users; }
    public Affiliate affiliate() { return affiliate; }
    public Billing billing() { return billing; }
    public Support support() { return support; }
    public Contacts contacts() { return contacts; }
    public Domains domains() { return domains; }
    public DnsManage dnsManage() { return dnsManage; }
    public Ssl ssl() { return ssl; }
    public Services services() { return services; }
    public Cart cart() { return cart; }
    public Dns dns() { return dns; }
    public Notifications notifications() { return notifications; }
    public Virtualizor virtualizor() { return virtualizor; }
    public CloudGpu cloudGpu() { return cloudGpu; }
    public CloudService cloudService() { return cloudService; }
    public NetworkServices networkServices() { return networkServices; }
    public ServiceStatus serviceStatus() { return serviceStatus; }
    public BareMetal bareMetal() { return bareMetal; }
    public Collocation collocation() { return collocation; }
    public Vcloudstack vcloudstack() { return vcloudstack; }
    public Hosting hosting() { return hosting; }
    public LocationV2 locationV2() { return locationV2; }
    public CloudInstance cloudInstance() { return cloudInstance; }
    public CloudVm cloudVm() { return cloudVm; }
    public AffiliatesAdvanced affiliatesAdvanced() { return affiliatesAdvanced; }
    public AiFactory aiFactory() { return aiFactory; }
    public PasskeyV2 passkeyV2() { return passkeyV2; }
    public EmailMfaV2 emailMfaV2() { return emailMfaV2; }
    public Vneidekyc vneidekyc() { return vneidekyc; }
    public WillExpired willExpired() { return willExpired; }
    public UrlShortener urlShortener() { return urlShortener; }
    public ProxmoxBackup proxmoxBackup() { return proxmoxBackup; }
    public CephS3 cephS3() { return cephS3; }
    public Pmg pmg() { return pmg; }
    public Proxmox proxmox() { return proxmox; }
    public Ipam ipam() { return ipam; }
    public Partner partner() { return partner; }
}

File: sdk/csharp/HiTechCloudClient.cs
namespace HiTechCloud.SDK;

public class HiTechCloudClient
{
    public const string ProductionUrl = "https://api.hitechcloud.vn";
    public const string OteUrl = "https://ote.hitechcloud.vn";
    public const string VERSION = "1.0.0";

    private readonly HttpClientWrapper _http;

    public Resources.Auth Auth { get; }
    public Resources.Users Users { get; }
    public Resources.Affiliate Affiliate { get; }
    public Resources.Billing Billing { get; }
    public Resources.Support Support { get; }
    public Resources.Contacts Contacts { get; }
    public Resources.Domains Domains { get; }
    public Resources.DnsManage DnsManage { get; }
    public Resources.Ssl Ssl { get; }
    public Resources.Services Services { get; }
    public Resources.Cart Cart { get; }
    public Resources.Dns Dns { get; }
    public Resources.Notifications Notifications { get; }
    public Resources.Virtualizor Virtualizor { get; }
    public Resources.CloudGpu CloudGpu { get; }
    public Resources.CloudService CloudService { get; }
    public Resources.NetworkServices NetworkServices { get; }
    public Resources.ServiceStatus ServiceStatus { get; }
    public Resources.BareMetal BareMetal { get; }
    public Resources.Collocation Collocation { get; }
    public Resources.Vcloudstack Vcloudstack { get; }
    public Resources.Hosting Hosting { get; }
    public Resources.LocationV2 LocationV2 { get; }
    public Resources.CloudInstance CloudInstance { get; }
    public Resources.CloudVm CloudVm { get; }
    public Resources.AffiliatesAdvanced AffiliatesAdvanced { get; }
    public Resources.AiFactory AiFactory { get; }
    public Resources.PasskeyV2 PasskeyV2 { get; }
    public Resources.EmailMfaV2 EmailMfaV2 { get; }
    public Resources.Vneidekyc Vneidekyc { get; }
    public Resources.WillExpired WillExpired { get; }
    public Resources.UrlShortener UrlShortener { get; }
    public Resources.ProxmoxBackup ProxmoxBackup { get; }
    public Resources.CephS3 CephS3 { get; }
    public Resources.Pmg Pmg { get; }
    public Resources.Proxmox Proxmox { get; }
    public Resources.Ipam Ipam { get; }
    public Resources.Partner Partner { get; }

    public HiTechCloudClient(string? token = null, string? baseUrl = null,
        bool sandbox = false, int timeout = 30, int maxRetries = 3)
    {
        var url = sandbox ? OteUrl : (baseUrl ?? ProductionUrl);
        _http = new HttpClientWrapper(url, timeout, maxRetries);

        if (token != null) _http.SetToken(token);

        Auth = new Resources.Auth(_http);
        Users = new Resources.Users(_http);
        Affiliate = new Resources.Affiliate(_http);
        Billing = new Resources.Billing(_http);
        Support = new Resources.Support(_http);
        Contacts = new Resources.Contacts(_http);
        Domains = new Resources.Domains(_http);
        DnsManage = new Resources.DnsManage(_http);
        Ssl = new Resources.Ssl(_http);
        Services = new Resources.Services(_http);
        Cart = new Resources.Cart(_http);
        Dns = new Resources.Dns(_http);
        Notifications = new Resources.Notifications(_http);
        Virtualizor = new Resources.Virtualizor(_http);
        CloudGpu = new Resources.CloudGpu(_http);
        CloudService = new Resources.CloudService(_http);
        NetworkServices = new Resources.NetworkServices(_http);
        ServiceStatus = new Resources.ServiceStatus(_http);
        BareMetal = new Resources.BareMetal(_http);
        Collocation = new Resources.Collocation(_http);
        Vcloudstack = new Resources.Vcloudstack(_http);
        Hosting = new Resources.Hosting(_http);
        LocationV2 = new Resources.LocationV2(_http);
        CloudInstance = new Resources.CloudInstance(_http);
        CloudVm = new Resources.CloudVm(_http);
        AffiliatesAdvanced = new Resources.AffiliatesAdvanced(_http);
        AiFactory = new Resources.AiFactory(_http);
        PasskeyV2 = new Resources.PasskeyV2(_http);
        EmailMfaV2 = new Resources.EmailMfaV2(_http);
        Vneidekyc = new Resources.Vneidekyc(_http);
        WillExpired = new Resources.WillExpired(_http);
        UrlShortener = new Resources.UrlShortener(_http);
        ProxmoxBackup = new Resources.ProxmoxBackup(_http);
        CephS3 = new Resources.CephS3(_http);
        Pmg = new Resources.Pmg(_http);
        Proxmox = new Resources.Proxmox(_http);
        Ipam = new Resources.Ipam(_http);
        Partner = new Resources.Partner(_http);
    }

    public async Task<Dictionary<string, object?>> LoginAsync(string email, string password)
    {
        var result = await Auth.LoginAsync(email, password);
        if (result.ContainsKey("token"))
            _http.SetToken(result["token"]?.ToString() ?? "");
        return result;
    }

    public async Task<Dictionary<string, object?>> LogoutAsync()
    {
        var result = await Auth.LogoutAsync();
        _http.ClearToken();
        return result;
    }
}
